mitodl/open-discussions
# support multiple departments

#### Pre-Flight checklist

#### What are the relevant tickets?
closes https://github.com/mitodl/open-discussions/issues/3408

#### What's this PR do?
This pr indexes all the course departments to elasticsearch when a course is crosslisted to multiple departments

#### How should this be manually tested?
Make sure the migration works backwards and forwards

Run
```
from course_catalog.api import *
from course_catalog.models import *

sync_ocw_courses(course_prefixes=["PROD/6/6.901/Fall_2005/6-901-inventions-and-patents-fall-2005/"], blocklist=[], force_overwrite=True, upload_to_s3=False)
```

Run ocw-hugo-themes with this branch for the index
Search for "Inventions and Patents"  
Verify that you see one  course and three departments

#### Any background context you want to provide?
I wasn't able to add the data migration in the migration because django tries to save the array field to postgres with `[` brackets instead of `{` when the `Course` model is accessed through `apps.get_model()` which results in a database error. Updating the data works fine when `Course` is accessed normally. I plan to run the following code from the command line to backfill the data manually:

```
        from course_catalog.models  import Course 
        from course_catalog.constants import PlatformType

        for course in Course.objects.filter(platform=PlatformType.ocw.value).iterator():
            course_runs = course.runs.filter(
                raw_json__isnull=False,
                published=True
            )

            for course_run in course_runs:
                for extra_course_number_json in (course_run.raw_json.get("extra_course_number") or []):
                    department_number=extra_course_number_json.get("linked_course_number_col").split('.')[0]
                    if department_number not in course.department:
                        course.department.append(department_number)
                        course.save()
```
